### PR TITLE
Automatically register @EndpointConverter-annotated Converter and GenericConverter beans with the ConversionService used by Actuator endpoints

### LINKED ISSUE #11013: ConversionService used for Actuator endpoints should be configurable
Because the `EndpointAutoConfiguration` `endpointOperationParameterMapper()` method is not annotated by `@ConditionalOnMissingBean(ParameterMapper.class)`, I had to resort to redundantly creating my own `ConversionServiceParameterMapper` (with my own, customized conversion service) with the `@Primary` annotation. This solved my problem, but now my code includes this line from `ConversionServiceParameterMapper`: `IsoOffsetDateTimeConverter.registerConverter(conversionService);` This isn't great because I now must remember to check the code in `ConversionServiceParameterMapper` when upgrading to newer versions of Spring Boot to make sure I don't need to configure my own `ConversionService` differently due to internal changes.  And, it's slightly sloppy because there's an extra `ParameterMapper` created for no reason.

I'd like for there to be a clean way to customize the Actuator-specific `ConversionService` used as the `ParameterMapper`.

### PR DESCRIPTION
**Pull request by chang-chao**
`Converter` and `GenericConverter ` beans annotated with `EndpointConverter` will be used in endpoint.
fixes #11013

### CODE REVIEW

**Comment by chang-chao** [comment]
Hi, @snicoll, would you please review this?

**Comment by philwebb** [comment]
@chang-chao @snicoll is at a conference at the moment and a lot of the team our away. We'll get to this as soon as we can.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/endpoint/EndpointAutoConfiguration.java [hunk 1]
```diff
@@ -16,30 +16,45 @@
...
 public class EndpointAutoConfiguration {
 
 	@Bean
 	@ConditionalOnMissingBean
-	public ParameterValueMapper endpointOperationParameterMapper() {
-		return new ConversionServiceParameterValueMapper();
+	public ParameterValueMapper endpointOperationParameterMapper(
+			ApplicationContext applicationContext) {
+		return new ConversionServiceParameterValueMapper(
+				new Factory(applicationContext.getAutowireCapableBeanFactory()).create());
```
Comment (nosan) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/endpoint/EndpointAutoConfiguration.java]: You are able to inject `Converter` and `GenericConverter` classes via `ObjectProvider`.
```java
	@Bean
	@ConditionalOnMissingBean
	public ParameterValueMapper endpointOperationParameterMapper(
			@EndpointConverter ObjectProvider<Converter<?, ?>> converters,
			@EndpointConverter ObjectProvider<GenericConverter> genericConverters) { 
		//
	}
```

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/endpoint/EndpointConverter.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+@Qualifier(EndpointConverter.VALUE)
+@Target({ ElementType.TYPE, ElementType.METHOD })
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+public @interface EndpointConverter {
+
+	/**
+	 * Concrete value for the {@link Qualifier @Qualifier}.
+	 */
+	String VALUE = "org.springframework.boot.actuate.autoconfigure.endpoint.EndpointConverter";
```
Comment (nosan) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/endpoint/EndpointConverter.java]: I believe that `@Qualifier` annotation can be used here. 

```java

@Target({ ElementType.FIELD, ElementType.METHOD, ElementType.PARAMETER, ElementType.TYPE, ElementType.ANNOTATION_TYPE })
@Retention(RetentionPolicy.RUNTIME)
@Documented
@Qualifier
public @interface EndpointConverter {

}

```

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/endpoint/EndpointAutoConfiguration.java (hunk 1 above)
Comment (wilkinsona) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/endpoint/EndpointAutoConfiguration.java]: Thanks, @nosan. I suspect the proposal is as it is as it has been inspired by `@ConfigurationPropertiesBinding` and `ConversionServiceDeducer`. The context here is a little bit different, however, as the configuration properties stuff is used from a bean post-processor. We can take a look at make some of the amendments suggested here as part of merging this.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/endpoint/EndpointConverter.java (hunk 2 above)
Comment (nosan) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/endpoint/EndpointConverter.java]: My bad, I've not seen that `@Qualifier` annotation has already been present :(

**Comment by snicoll** [comment]
@chang-chao thank you, this is now merged in `master`.

**Pull request closed without merging**